vars/cloudFoundryDeploy.groovy

```diff
@@ -0,0 +1,166 @@
+}
+
+def deployCfNative (config) {
+    withCredentials([usernamePassword(
+        credentialsId: config.cloudFoundry.credentialsId,
+        passwordVariable: 'password',
+        usernameVariable: 'username'
+    )]) {
+        def deployCommand = 'push'
+        if (config.deployType == 'blue-green') {
+            deployCommand = 'blue-green-deploy'
+        } else {
+            config.smokeTest = ''
+        }
+
+        // check if appName is available
+        if (config.cloudFoundry.appName == null || config.cloudFoundry.appName == '') {
+            if (fileExists(config.cloudFoundry.manifest)) {
+                def manifest = readYaml file: config.cloudFoundry.manifest
+                if (!manifest || !manifest.applications || !manifest.applications[0].name)
+                    error "[${STEP_NAME}] ERROR: No appName available in manifest ${config.cloudFoundry.manifest}."
+
+            } else {
+                error "[${STEP_NAME}] ERROR: No manifest file ${config.cloudFoundry.manifest} found."
+            }
+        }
+
+        sh """#!/bin/bash
+            set +x
+            export HOME=${config.dockerWorkspace}
+            cf login -u \"${username}\" -p '${password}' -a ${config.cloudFoundry.apiEndpoint} -o \"${config.cloudFoundry.org}\" -s \"${config.cloudFoundry.space}\"
+            cf plugins
+            cf ${deployCommand} ${config.cloudFoundry.appName?"\"${config.cloudFoundry.appName}\"":''} -f \"${config.cloudFoundry.manifest}\" ${config.smokeTest}"""
+        def retVal = sh script: "cf app \"${config.cloudFoundry.appName}-old\"", returnStatus: true
+        if (retVal == 0) {
+            sh "cf delete \"${config.cloudFoundry.appName}-old\" -r -f"
+        }
+        sh "cf logout"
+    }
+}
+
+def deployMta (config) {
+    if (config.mtaExtensionDescriptor == null) config.mtaExtensionDescriptor = ''
+    if (!config.mtaExtensionDescriptor.isEmpty() && !config.mtaExtensionDescriptor.startsWith('-e ')) config.mtaExtensionDescriptor = "-e ${config.mtaExtensionDescriptor}"
+
+    def deployCommand = 'deploy'
+    if (config.deployType == 'blue-green')
+        deployCommand = 'bg-deploy'
+
+    withCredentials([usernamePassword(
+        credentialsId: config.cloudFoundry.credentialsId,
+        passwordVariable: 'password',
+        usernameVariable: 'username'
+    )]) {
+        echo "[${STEP_NAME}] Deploying MTA (${config.mtaPath}) with following parameters: ${config.mtaExtensionDescriptor} ${config.mtaDeployParameters}"
+        sh """#!/bin/bash
+            export HOME=${config.dockerWorkspace}
+            set +x
+            cf api ${config.cloudFoundry.apiEndpoint}
+            cf login -u ${username} -p '${password}' -a ${config.cloudFoundry.apiEndpoint} -o \"${config.cloudFoundry.org}\" -s \"${config.cloudFoundry.space}\"
+            cf plugins
+            cf ${deployCommand} ${config.mtaPath} ${config.mtaDeployParameters} ${config.mtaExtensionDescriptor}"""
+        sh "cf logout"
```

Comment:
what is the reason for having ```cf logout``` in an other shell instead of having it the previous shell? In case the reason is: 'we would like to ensure that the logout happens in any case' we could also use good old [trap](https://www.google.com/search?q=bash+trap). See also above

Reply:
no real reason. It just works fine this way ...